### ISSUE TITLE
[BUG] Playwright report doesn't have test report attachments 

### ISSUE THREAD

**Issue by akhilkanjarla** [microsoft/playwright#28348]
When a Playwright test encounters a timeout during the **afterEach** step, we observe that Playwright result attachments such as screenshots and videos fail to be properly linked in the Playwright report. This issue persists even when the screenshots and videos are present in the test-results folder

### System info
- Playwright Version: [v1.32.0]
- Operating System: [All, Windows 11, Ubuntu 20, macOS 13.2, etc.]
- Browser: [All, Chromium, Firefox, WebKit]
- Other info:

### Source code

- [ x ] I provided exact source code that allows reproducing the issue locally.
 (can't share the exact code as it's internal to our org)

Playwright test error
<img width="1336" alt="image" src="https://github.com/microsoft/playwright/assets/88868745/2867b039-7523-4550-9d8d-7abb90886170">

test-results.json

<img width="1336" alt="image" src="https://github.com/microsoft/playwright/assets/88868745/fcbd65ad-c9f9-4ecd-a98b-7946584e14fd">

test-results folder

<img width="1402" alt="image" src="https://github.com/microsoft/playwright/assets/88868745/6d546a5c-c1c2-4bf0-b1d1-99edb8985cb7">

**Config file**

```js
// playwright.config.ts
reporter: [["list"], ["json", { outputFile: "test-results.json" }], ['html', { open: 'never' }]],
  /* Shared settings for all the projects below. See https://playwright.dev/docs/api/class-testoptions. */
  use: {
    /* Maximum time each action such as `click()` can take. Defaults to 0 (no limit). */
    actionTimeout: 0,
    /* Base URL to use in actions like `await page.goto('/')`. */
    // baseURL: 'http://localhost:3000',

    /* Collect trace when retrying the failed test. See https://playwright.dev/docs/trace-viewer */
    screenshot: "only-on-failure",
    video: "on-first-retry",
  },

```

**Test file (self-contained)**

```js
test.beforeEach(async ({page}) => {
  <dummy logic>
});

test.afterEach(async ({page}) => {
    <dummy logic to get  timedout here>
});

test('automationops-governance-main-flow', async ({browser, page}) => {
    <dummy logic>
});
```

**Steps**
- Add any dummy logic as needed so it gets timedout in afterEach step (can't share exact code as it's internal to our org)

**Expected**

Test results should be linked in the test reports like test-result.json , html report 

**Actual**

Test attachments are coming empty in the test reports.

**Comment by mxschmitt** [comment]
Investigation: Looks like we receive the 'done' event before we have teared down the fixtures.

**Comment by mxschmitt** [comment]
Feel free to re-file if you still encounter any issues.

**Comment by mmun** [comment]
@mxschmitt We've been running into this issue at work using the latest version of Playwright (1.40.1).

I've created a minimal reproduction here: https://github.com/mmun/playwright-issue-28348

I'm not sure if its related, but in addition to the issue described here I also noticed that `afterEach` does not respect the `test.setTimeout` defined in the test, and instead always uses the default. This behavior is in contrast to the behavior described in this comment: https://github.com/microsoft/playwright/issues/24285#issuecomment-1647857959.

In any case, let me know if you'd prefer new issues to be opened.
